### PR TITLE
Fix MULTIDRIVEN with task and default driver (#4045)

### PR DESCRIPTION
**Pull request by em2machine**
Per the end of #4045:

Treat “task definition” as not a process (don’t set the who/where the writer was), so figure out how to gate this appropriately.
Treat “call site” as part of the caller’s process (iterate/descend through the statements of the task setting who/where writer as appropriate).

I created V3UndrivenCapture to hold the state, and an enable to disable the new behaviour if desired.  Added a bunch of tests and alot of comments in V3UndrivenCapture.h itself.

### CODE REVIEW

File: test_regress/t/t_lint_multidriven_taskcall_bad.v [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// DESCRIPTION: Verilator: Verilog Test module
```
Comment (wsnyder) [inline test_regress/t/t_lint_multidriven_taskcall_bad.v:1]: Please rename t_lint_multidriven_taskcall_bad.  (So near other similar tests)

File: test_regress/t/t_lint_taskcall_multidriven_bad.v [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+//
+// This file ONLY is placed into the Public Domain, for any use,
+// without warranty.
+// SPDX-License-Identifier: CC0-1.0
+
+module t (
+    input  logic sel
+  , input  logic sel2
+  , input  logic d
+  , output logic out
```
Comment (wsnyder) [inline test_regress/t/t_lint_taskcall_multidriven_bad.v]: Commas at end, likewise for in instance pins.

```suggestion
    input  logic sel,
    input  logic sel2,
    input  logic d,
    output logic out
```

File: test_regress/t/t_multidriven_class0.v [hunk 3]
```diff
@@ -0,0 +1,72 @@
+// DESCRIPTION: Verilator: Verilog Test module
```
Comment (wsnyder) [inline test_regress/t/t_multidriven_class0.v]: You're adding a lot of small tests where a lot of the code is just setup/in common, can several of these cases be combined into larger tests?

File: src/V3Undriven.cpp [hunk 4]
```diff
@@ -51,6 +52,9 @@ class UndrivenVarEntry final {
     const FileLine* m_nodeFileLinep = nullptr;  // File line of varref if driven, else nullptr
     bool m_underGen = false;  // Under a generate
 
+    const AstNode* m_callNodep = nullptr;  // call node if driven via writeSummary, else nullptr
+    const FileLine* m_callFileLinep = nullptr;  // file line of call node if driven via summary
```
Comment (wsnyder) [inline src/V3Undriven.cpp]: Capitalize comments like sentences please.  (Everywhere)

```suggestion
    const AstNode* m_callNodep = nullptr;  // Call node if driven via writeSummary, else nullptr
    const FileLine* m_callFileLinep = nullptr;  // Fileline of call node if driven via summary
```

File: src/V3Undriven.cpp [hunk 5]
```diff
@@ -278,6 +282,16 @@ class UndrivenVarEntry final {
...
+
+    void drivenViaCall(const AstNode* nodep, const FileLine* fileLinep) {
+        drivenWhole();
+        if (!m_callNodep) {
+            m_callNodep = nodep;
+            m_callFileLinep = fileLinep;
+        }
+    }
+    const AstNode* getCallNodep() const { return m_callNodep; }
+    const FileLine* getCallFileLinep() const { return m_callFileLinep; }
```
Comment (wsnyder) [inline src/V3Undriven.cpp]: For historical reasons it's Fileline not FileLine although agree the latter would have made more sense.

But do we even need the callFilelinep?  Given how this function is called you can just use callNodep->fileline() to get it.

Should this function get a NodeFTaskRef instead of Node?

```suggestion
    const AstNode* callNodep() const { return m_callNodep; }
    const FileLine* callFilelinep() const { return m_callFilelinep; }
```

File: src/V3Undriven.cpp [hunk 6]
```diff
@@ -304,6 +318,9 @@ class UndrivenVisitor final : public VNVisitorConst {
     const AstAlways* m_alwaysp = nullptr;  // Current always of either type
     const AstAlways* m_alwaysCombp = nullptr;  // Current always if combo, otherwise nullptr
 
+    V3UndrivenCapture* const m_capturep = nullptr;
+    const bool m_enableWriteSummary = false;
```
Comment (wsnyder) [inline src/V3Undriven.cpp]: Comments on members please.

File: src/V3Undriven.cpp [hunk 7]
```diff
@@ -523,10 +554,43 @@ class UndrivenVisitor final : public VNVisitorConst {
...
+        //    return !(m_taskp && !m_alwaysp && !m_inContAssign && !m_inInitialStatic &&
+        //    !m_inBBox);
+        //};
+
+        const auto inExecutedContext = [this]() {
+            return !(m_taskp && !m_alwaysp && !m_inContAssign && !m_inInitialStatic && !m_inBBox
+                     && !m_taskp->dpiExport());
+        };
+
+        if (!inExecutedContext()) { return; }
```
Comment (wsnyder) [inline src/V3Undriven.cpp]:
```suggestion
        if (!inExecutedContext()) return;
```

File: src/V3UndrivenCapture.h [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        // 'write through write' writeSummary for the given task/function.  Meaning ultimately
+        // everything that this function/task writes to.
+        std::vector<Var> writeSummary;
+        // state for writeSummary computation.
+        State state = State::UNVISITED;
+    };
+
+    // Enable writeSummary computation.  If disabled, then the existing V3Undriven behaviour is
+    // used.
+    static bool enableWriteSummary;
```
Comment (wsnyder) [inline src/V3UndrivenCapture.h]: Is this just for current debug, or will be removed?  Why do we need both behaviors ultimately?

File: src/V3UndrivenCapture.h [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+#include <unordered_map>
+#include <vector>
+
+class AstNetlist;
+
+class V3UndrivenCapture final {
+public:
+    using FTask = const AstNodeFTask*;
+    using Var = AstVar*;
```
Comment (wsnyder) [inline src/V3UndrivenCapture.h]: Don't do this please, expand where used to be explicit.

File: src/V3UndrivenCapture.cpp [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+//
+//*************************************************************************
+//
+// Copyright 2003-2025 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
```
Comment (wsnyder) [inline src/V3UndrivenCapture.cpp:15]:
```suggestion
//
//*************************************************************************
//
// ... add summary about what this does/how works(like in other .cpp files)
//
//*************************************************************************
```

File: src/V3UndrivenCapture.cpp [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+#include "V3Error.h"
+#include "V3Global.h"
+
+#include <algorithm>
+
+VL_DEFINE_DEBUG_FUNCTIONS;
+
+namespace {
+
+constexpr int DBG = 9;
```
Comment (wsnyder) [inline src/V3UndrivenCapture.cpp]: Hardcode the 9, and use `--debugi-V3UndrivenCapture 9` when debugging this.

File: src/V3UndrivenCapture.cpp [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    vec.erase(std::unique(vec.begin(), vec.end()), vec.end());
+}
+
+V3UndrivenCapture::V3UndrivenCapture(AstNetlist* netlistp) {
+    gather(netlistp);
+
+    // Normalize direct lists
+    for (auto& kv : m_info) {
+        sortUniqueVars(kv.second.directWrites);
+        sortUniqueFTasks(kv.second.callees);
```
Comment (wsnyder) [inline src/V3UndrivenCapture.cpp]: Why do we need to sort?  Can we avoid by relying on the container sorting as we insert, or only sort when have an error?  I'm worried about performance.

Comment (em2machine) [reply src/V3UndrivenCapture.cpp]: Yes, I agree, I had some mental gymnastics here on practicality versus complexity.  The primary reason for these is to dedup both sets, but yes this is done at the end and so there is a storage penalty for sure and performance seems like it can be impacted in canonical worst case - although I don't have any tests to prove.  I could change to do on insert, but this means keeping around an unordered_map<> (temporary storage) for both sets and then save the final vector.  The simplest to change without ripping out much plumbing would be for callees as I think thats where duplicates are the most likely.  So we could:
a) leave as is for now until / unless we see a significant performance degradation and then change
b) leave as is for now and I start working on the temporary filter on the callee side plumbing as a followup
c) wait and keep working on the filter for both direct writes and callee sites.  
Anecdotally on a large SOC with tons of AXI interfaces where I had MULTIDRIVEN disabled it now Verilates faster then before - not sure I can justify why this is yet...  (and of course correlation doesn't necessarily equal causation :) ).

File: test_regress/t/t_lint_multidriven_taskcall_bad.v (hunk 1 above)
Comment (em2machine) [reply test_regress/t/t_lint_multidriven_taskcall_bad.v:1]: Done.  Should have noticed this.

File: test_regress/t/t_lint_taskcall_multidriven_bad.v (hunk 2 above)
Comment (em2machine) [reply test_regress/t/t_lint_taskcall_multidriven_bad.v]: This is a habit thats going to be hard to break. :). Fixed.

File: test_regress/t/t_multidriven_class0.v (hunk 3 above)
Comment (em2machine) [reply test_regress/t/t_multidriven_class0.v]: I'll be committing an aggregated t_multidriven_simple.v test.  Take a look and let me know if this is what you had in mind.  They all have some subtle differences (and it makes it easier for me to look at the trees, etc), but I can do the same with the class and interface variants - but it does amount to pulling the individual test modules into a single test and running and checking in parallel.

File: src/V3UndrivenCapture.h (hunk 8 above)
Comment (em2machine) [reply src/V3UndrivenCapture.h]: Ahem. :). Its a ripcord that allows for quick disable.  I'm paranoid. :)  I can remove if there is heartburn.

File: src/V3UndrivenCapture.h (hunk 9 above)
Comment (em2machine) [reply src/V3UndrivenCapture.h]: Fixed.  Some laziness on my part.

File: src/V3Undriven.cpp (hunk 5 above)
Comment (em2machine) [reply src/V3Undriven.cpp]: Updated.  Changed to AstNodeFTaskRef as this is all that ever does get passed to drivenViaCall().

File: test_regress/t/t_multidriven_class0.v (hunk 3 above)
Comment (wsnyder) [reply test_regress/t/t_multidriven_class0.v]: To the extent it makes sense effort wise.  It's a lot faster for the regressions if tests are fewer and larger, but in the grand context it's not super critical if a lot of work.

File: src/V3UndrivenCapture.h (hunk 8 above)
Comment (wsnyder) [reply src/V3UndrivenCapture.h]: As this is only a warning that can be disabled I'd prefer we just turn it on always.

File: src/V3UndrivenCapture.cpp (hunk 12 above)
Comment (wsnyder) [reply src/V3UndrivenCapture.cpp]: I'd prefer you work out some alternative, up to you which.  We don't have many tests that check the crazy high-count edge cases, so lack of something breaking doesn't well prove performance won't give a surprise in the field.  We tend to add them only as something breaks, but I certainly welcome reasonable edge-case tests to improve this... the _deep_ tests have some examples.

Comment (em2machine) [reply src/V3UndrivenCapture.cpp]: Ok, I will add two tier filtering.  Once on insert for both call chains and for direct writes, and then when consolidating.  Once summaries are computed I will make sure the filters are empty (storage).  I pulled sorting as the only reason I could think of needing it would have been determinism for debug scenarios, etc.

File: src/V3UndrivenCapture.h (hunk 8 above)
Comment (em2machine) [reply src/V3UndrivenCapture.h]: Ok, will remove.

File: test_regress/t/t_multidriven_class0.v (hunk 3 above)
Comment (em2machine) [reply test_regress/t/t_multidriven_class0.v]: Will aggregate the others.

**Comment by em2machine** [comment]
@wsnyder I think this should be good to go.  Cheers, -e-

File: src/V3UndrivenCapture.h [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+//*************************************************************************
+
+//*************************************************************************
+//
+// Capture task/function write summaries for multidriven checks.
+// Per-task/function capture info keyed by resolved AstNodeFTask* identity (FTask = function or
+// task).  This is our 'graph' of tasks/functions.  Each node has a list of direct callees and
+// a list of variables written in the function body.  There are methods to dedup after walking the
+// tree. V3Undriven then uses the writeSummary for multidriven checks - i.e. it treats writes (side
+// effects) inside subroutines as part of the caller's process.
```
Comment (wsnyder) [inline src/V3UndrivenCapture.h]: Please fill-paragraph to column 90 or so.  Not sure why autoformatter didn't.

File: src/V3UndrivenCapture.h [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+class AstNetlist;
+
+class V3UndrivenCapture final {
+public:
+    // DFS computation state for writeSummary propagation.
+    // UNVISITED: write summary not computed yet
+    // VISITING: currently computing on the call stack - used to detect cycles
+    // DONE: write summary computed
+    enum class State : uint8_t { UNVISITED, VISITING, DONE };
```
Comment (wsnyder) [inline src/V3UndrivenCapture.h]:
```suggestion
    enum class State : uint8_t { UNVISITED,  // Write summary not computed yet
     VISITING,  // Currently computing on the call stack - used to detect cycles
      DONE  // Write summary computed
       };
```

File: src/V3UndrivenCapture.h [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+        // Direct resolved callees from this task/function body (iteration order)
+        std::vector<const AstNodeFTask*> callees;
+        // 'Write through write' writeSummary for the given task/function.  Meaning ultimately
+        // everything that this function/task writes to.
+        std::vector<AstVar*> writeSummary;
+        // State for writeSummary computation.
+        State state = State::UNVISITED;
+        // This is used to test whether weve already recorded a callee.  Used to 'filter' on insert
+        // versus sorting at the end.
+        std::unordered_set<const AstNodeFTask*> calleesSet;
```
Comment (wsnyder) [inline src/V3UndrivenCapture.h]:
```suggestion
        // Test if already recorded a callee.  Used to 'filter' on insert
        // versus sorting at the end.
        std::unordered_set<const AstNodeFTask*> calleesSet;
```

File: src/V3UndrivenCapture.h [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+        // This is used to test whether weve already recorded a direct write.  Used to 'filter' on
+        // insert versus sorting at the end.
+        std::unordered_set<AstVar*> directWritesSet;
```
Comment (wsnyder) [inline src/V3UndrivenCapture.h]:
```suggestion
        // Test if already recorded a direct write.  Used to 'filter' on
        // insert versus sorting at the end.
        std::unordered_set<AstVar*> directWritesSet;
```

File: src/V3UndrivenCapture.h [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+    };
+
+private:
+    // Per-task/function capture info keyed by resolved AstNodeFTask* identity (FTask = function or
+    // task).  This is our 'graph' of tasks/functions.  Each node has a list of direct callees and
```
Comment (wsnyder) [inline src/V3UndrivenCapture.h]: Nit avoid "our" etc assuming reader is the computer/programmer.

```suggestion
    // task).  The 'graph' of tasks/functions.  Each node has a list of direct callees and
```

File: src/V3Undriven.cpp [hunk 18]
```diff
@@ -51,6 +52,9 @@ class UndrivenVarEntry final {
     const FileLine* m_nodeFileLinep = nullptr;  // File line of varref if driven, else nullptr
     bool m_underGen = false;  // Under a generate
 
+    const AstNodeFTaskRef* m_callNodep
+        = nullptr;  // Call node if driven via writeSummary, else nullptr
```
Comment (wsnyder) [inline src/V3Undriven.cpp]:
```suggestion
    const AstNodeFTaskRef* m_callNodep = nullptr;  // Call node if driven via writeSummary

File: src/V3Undriven.cpp [hunk 19]
```diff
@@ -278,6 +282,12 @@ class UndrivenVarEntry final {
             }
         }
     }
+
+    void drivenViaCall(const AstNodeFTaskRef* nodep) {
+        drivenWhole();
+        if (!m_callNodep) { m_callNodep = nodep; }
```
Comment (wsnyder) [inline src/V3Undriven.cpp]:
```suggestion
        if (!m_callNodep) m_callNodep = nodep;
```

File: src/V3Undriven.cpp [hunk 20]
```diff
@@ -523,10 +548,39 @@ class UndrivenVisitor final : public VNVisitorConst {
...
+        // Do not apply writeSummary at calls inside a task definition, or they will look like
+        // independent drivers (phantom MULTIDRIVEN). Did the lambda on purpose - lessen chance of
+        // screwup in future edits.
+
+        const auto inExecutedContext = [this]() {
+            return !(m_taskp && !m_alwaysp && !m_inContAssign && !m_inInitialStatic && !m_inBBox
+                     && !m_taskp->dpiExport());
+        };
+
+        if (!inExecutedContext()) return;
```
Comment (wsnyder) [inline src/V3Undriven.cpp]: No need for Lambda unless I'm missing something.

```suggestion
        const bool inExecutedContext = 
             !(m_taskp && !m_alwaysp && !m_inContAssign && !m_inInitialStatic && !m_inBBox
                     && !m_taskp->dpiExport());
        if (!inExecutedContext) return;
```

File: src/V3UndrivenCapture.cpp [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+//*************************************************************************
+
+#include "V3UndrivenCapture.h"
+
+#include "V3Error.h"
+#include "V3Global.h"
+
+VL_DEFINE_DEBUG_FUNCTIONS;
+
+namespace {
```
Comment (wsnyder) [inline src/V3UndrivenCapture.cpp:24]: We generally make classes instead for these sort of things e.g. V3UndrivenImp

File: src/V3UndrivenCapture.cpp [hunk 22]
```diff
@@ -0,0 +1,212 @@
...
+
+struct Stats final {
+    uint64_t ftasks = 0;
+    uint64_t varWrites = 0;
+    uint64_t callEdges = 0;
```
Comment (wsnyder) [inline src/V3UndrivenCapture.cpp]: I think can be moved into the Visitor if print on destruction.  Should these be V3Stats instead so go into __stats.txt, or you think only really needed for debug?   Not sure they are needed at all once you have algorithm solid.

File: src/V3UndrivenCapture.cpp [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+} g_stats;
+
+static std::string taskNameQ(const AstNodeFTask* taskp) {
+    if (!taskp) return "<null>";
+    return taskp->prettyNameQ();
+}
+
+class CaptureVisitor final : public VNVisitorConst {
+    V3UndrivenCapture& m_cap;
+    const AstNodeFTask* m_curTaskp = nullptr;
```
Comment (wsnyder) [inline src/V3UndrivenCapture.cpp]:
```suggestion
    const AstNodeFTask* m_curTaskp = nullptr;  // Current task
```

File: src/V3UndrivenCapture.cpp [hunk 24]
```diff
@@ -0,0 +1,212 @@
...
+
+    static void iterateListConst(VNVisitorConst& v, AstNode* nodep) {
+        for (AstNode* np = nodep; np; np = np->nextp()) np->accept(v);
+    }
```
Comment (wsnyder) [inline src/V3UndrivenCapture.cpp]: Use iterateAndNextConstNull

File: src/V3UndrivenCapture.cpp (hunk 22 above)
Comment (em2machine) [reply src/V3UndrivenCapture.cpp]: These can be removed.  They were there for my debug.

File: src/V3Undriven.cpp (hunk 20 above)
Comment (em2machine) [reply src/V3Undriven.cpp]: Fixed.  Was a holdover as I was thinking I was going to need this in multiple places.

File: src/V3UndrivenCapture.cpp (hunk 24 above)
Comment (em2machine) [reply src/V3UndrivenCapture.cpp]: Done

File: src/V3UndrivenCapture.cpp (hunk 21 above)
Comment (em2machine) [reply src/V3UndrivenCapture.cpp:24]: Created a small CaptureUtils container to hold the static function.

**Comment by em2machine** [comment]
Addressing the coverage now.  The missing lines were due to internal debug functionality.  Removing those bits and then will local rebase and push to make sure everything still works.

**Comment by em2machine** [comment]
@wsnyder  ok, I think I'm done.  The last few lines coverage wise look related to calling UINFO() debug functionality...
